contacts/tests.py

```diff
@@ -0,0 +1,170 @@
+        # else validation fails
+        self.assertEqual(
+            test_case.date_start.replace(tzinfo=None),
+            datetime.fromisoformat(data.get("timestamp")),
+        )
+
+        # check if enddate is calculated correctly
+        self.assertEqual(
+            test_case.date_end.replace(tzinfo=None),
+            datetime.fromisoformat(data.get("timestamp")) + timedelta(days=14),
+        )
+
+    def test_case_active(self):
+        url = reverse("contacts:rest_confirmed_contact")
+
+        # data_later -> is_active True
+        data_later = self.generate_data(hours=10)
+        # data_earlier -> is_active False
+        data_earlier = self.generate_data(hours=12)
+
+        r = self.client.post(url, data_later)
+        self.assertEquals(r.status_code, 201)
+
+        r2 = self.client.post(url, data=data_earlier)
+        self.assertEquals(r2.status_code, 201)
+
+        logger.info("Submitted two cases")
+
+        case_earlier = Case.objects.filter(
+            external_id=data_earlier.get("external_id")
+        ).first()
+        case_later = Case.objects.filter(
+            external_id=data_later.get("external_id")
+        ).first()
+
+        self.assertIsNotNone(case_earlier)
+        self.assertIsNotNone(case_later)
+
+        test_contact = Contact.objects.filter(msisdn=data_earlier.get("msisdn")).first()
+
+        self.assertIsNotNone(test_contact)
+
+        self.assertEqual(case_earlier.contact, test_contact)
+        self.assertEqual(case_earlier.contact, case_later.contact)
+
+        self.assertFalse(case_earlier.is_active)
+        self.assertTrue(case_later.is_active)
+
+        logger.info("Finished testing two cases")
+
+    def test_bad_request_case(self):
+        url = reverse("contacts:rest_confirmed_contact")
+
+        valid_time = (datetime.now() - timedelta(hours=72)).isoformat()
+
+        # data_later -> is_active True
+        data_malformed = [
+            # invalid phone number
+            {
+                "msisdn": "qwerty",
+                "external_id": uuid4().hex,
+                "timestamp": valid_time,
+            },  # noqa: E231, E261, E501
+            # invalid uuid
+            {
+                "msisdn": "+27820001001",
+                "external_id": None,
+                "timestamp": valid_time,
+            },  # noqa: E231, E261, E501
+            # invalid timestamp (future)
+            {
+                "msisdn": "+27820001001",
+                "external_id": uuid4().hex,
+                "timestamp": (datetime.now() + timedelta(hours=72)).isoformat(),
+            },
+            # invalid timestamp (expired)
+            {
+                "msisdn": "+27820001001",
+                "external_id": uuid4().hex,
+                "timestamp": (datetime.now() - timedelta(days=15)).isoformat(),
+            },
+        ]
+
+        # first two cases will raise 200_OK, other two 400_BAD_REQUEST
+        expected_responses = [200, 200, 400, 400]
```

Comment:
I would've expected 400 responses for all of these? In all the cases, the user of the API has supplied invalid data, so we should tell them that by responding with a 400. By responding with a 200 we're saying that we've handled it successfully, when we haven't handled it, we're rejecting it.

Reply:
Rewrote tests according to new changes, now all errors are expected to return 200.